libs/core-ui/src/lib/DatasetCohort.ts

```diff
@@ -0,0 +1,236 @@
+  ICategoricalRange,
+  INumericRange,
+  RangeTypes
+} from "@responsible-ai/mlchartlib";
+
+import { ErrorCohortStats, MetricCohortStats } from "./Cohort/CohortStats";
+import { CohortSource, Metrics } from "./Cohort/Constants";
+import { DatasetCohortColumns } from "./DatasetCohortColumns";
+import { IDataset } from "./Interfaces/IDataset";
+import { ModelTypes } from "./Interfaces/IExplanationContext";
+import { FilterMethods, IFilter } from "./Interfaces/IFilter";
+import { getPropertyValues } from "./util/datasetUtils/getPropertyValues";
+import { IsBinary, IsMulticlass } from "./util/ExplanationUtils";
+import { MulticlassClassificationEnum } from "./util/JointDatasetUtils";
+
+export class DatasetCohort {
+  public selectedIndexes: number[] = [];
+  public cohortStats: MetricCohortStats;
+  public constructor(
+    public name: string,
+    public dataset: IDataset,
+    public filters: IFilter[] = [],
+    public modelTypes?: ModelTypes,
+    private featureRanges?: {
+      [key: string]: INumericRange | ICategoricalRange;
+    },
+    public source: CohortSource = CohortSource.None,
+    public isTemporary: boolean = false,
+    cohortStats: MetricCohortStats | undefined = undefined
+  ) {
+    this.name = name;
+    this.selectedIndexes = this.applyFilters();
+    if (cohortStats) {
+      this.cohortStats = cohortStats;
+    } else {
+      this.cohortStats = this.updateStatsFromData();
+    }
+  }
+
+  private applyFilters(): number[] {
+    const indexes = [];
+    const dataDict = this.getDataDict(this.modelTypes);
+    for (const [index, row] of dataDict.entries()) {
+      if (this.filterRow(row, this.filters)) {
+        indexes.push(index);
+      }
+    }
+    return indexes;
+  }
+
+  private getDataDict(
+    modelType?: ModelTypes
+  ): Array<{ [key: string]: unknown }> {
+    const dataDict = Array.from({ length: this.dataset.features.length }).map(
+      (_, index) => {
+        const dict = {};
+        dict[DatasetCohortColumns.Index] = index;
+        return dict;
+      }
+    );
+    this.dataset.features.forEach((row, index) => {
+      row.forEach((val, colIndex) => {
+        const featureName = this.dataset.feature_names[colIndex];
+        dataDict[index][featureName] = val;
+      });
+    });
+    this.dataset.true_y.forEach((val, index) => {
+      if (Array.isArray(val)) {
+        val.forEach((subVal, subIndex) => {
+          dataDict[index][DatasetCohortColumns.TrueY + subIndex.toString()] =
+            subVal;
+        });
+      } else {
+        dataDict[index][DatasetCohortColumns.TrueY] = val;
+      }
+    });
+    this.dataset.predicted_y?.forEach((val, index) => {
+      if (Array.isArray(val)) {
+        val.forEach((subVal, subIndex) => {
+          dataDict[index][
+            DatasetCohortColumns.PredictedY + subIndex.toString()
+          ] = subVal;
+        });
+      } else {
+        dataDict[index][DatasetCohortColumns.PredictedY] = val;
+      }
+    });
+    // set up errors
```

Comment:
do we consider multi label classification scenario?

Reply:
is line 110 what you are asking?

Reply:
for multiclass error is line 110
for true Y and predict Y multi label scenario, it will be line 74 and line 85,
```
val.forEach((subVal, subIndex) => {
          dataDict[index][
            DatasetCohortColumns.PredictedY + subIndex.toString()
          ] = subVal;
```